### PR TITLE
Ensure map decoding signals fields being consumed

### PR DESCRIPTION
**Pull request by thomasvl**
This isn't meant to currently be a PR, I'm opening it for discussion because I'm a little confused.

### CODE REVIEW

File: Sources/SwiftProtobuf/BinaryDecoder.swift [hunk 1]
```diff
@@ -895,6 +895,7 @@ internal struct BinaryDecoder: Decoder {
 
         if let k = k, let v = v {
             value[k] = v
+            consumed = true
```
Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: @tbkka When looking at moving unknowns into the base Message, and always haven't them, so Maps tests started to failing, looking a little it seems like maps never say they consumed their input for the might have been ended up in both the unknowns and the real map field.  Does this fix seem correct?

File: Tests/SwiftProtobufTests/Test_Map.swift [hunk 2]
```diff
@@ -68,6 +68,7 @@ class Test_Map: XCTestCase, PBTestHelpers {
         assertDecodeSucceeds([10, 4, 8, 1, 16, 2]) {
             $0.mapInt32Int32 == [1: 2]
         }
+        // TODO: Why should this fail?  It appears to actually be valid.
```
Comment (thomasvl) [inline Tests/SwiftProtobufTests/Test_Map.swift]: After making the consumed changed, this test starts failing.  The comment says it is an invalid bad wire type, but I think it actually is a field 1 startGroup.  Where it gets interesting is I don't see our unknown handling property dealing with this.  Our generated `messageDecode` is:

```swift
  mutating func _protoc_generated_decodeMessage<D: SwiftProtobuf.Decoder>(decoder: inout D) throws {
    try withExtendedLifetime(_storage) { (_storage: _StorageClass) in
      while let fieldNumber = try decoder.nextFieldNumber() {
        try decodeField(decoder: &decoder, fieldNumber: fieldNumber)
      }
    }
  }
```

But how do groups factor into this.  Isn't this going to say it is field 1, but never really capture that it it was a startGroup and there *has* to be an endGroup marker and all sub data could into this messages unknowns.

File: Sources/SwiftProtobuf/BinaryDecoder.swift (hunk 1 above)
Comment (tbkka) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: I believe your proposed fix here is correct.

I'm a little dismayed that our tests did not catch this earlier; I should go back and figure out what tests we're missing.

As a minor matter of style, I would put this statement outside of this conditional, and might even pull it up to just after `getFieldBodyBytes` above.  When you put it here, it makes `consumed` look like it's flagging decoding success when it's really intended to just mark that the field has been inspected.

Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: So in addition to the test I tagged in the second commit in this PR putting it any other place than here causes some other things to fail.  I need to go back to the C++ code to figure out what happens if a map is missing the key or value on the wire.  As we're currently written, I don't think we'll parse it, but when I did maps for C++ I almost thing the C++ would end up using the default value.  The really crappy part is for a proto2 syntax file, the map can be to enums, so the value could be non zero making it harder to hardcode.

Comment (tbkka) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: The default initializer for an enum should DTRT for that case.

Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: Right now the code makes both Key and Value optional, so unless we decode them we don't assign into the map.  I'll keep digging for the C++ side.

Comment (thomasvl) [inline Sources/SwiftProtobuf/BinaryDecoder.swift]: Ok, found some old email to the protobuf-team confirming the default value part.  I'll look at fixing that up in another PR.

File: Tests/SwiftProtobufTests/Test_Map.swift (hunk 2 above)
Comment (thomasvl) [inline Tests/SwiftProtobufTests/Test_Map.swift]: @tbkka I'll update this PR to disable this test for now, but do you recall the intent here and how were groups supposed to be handled with unknowns, I'm I reading this correctly as not handing correctly?

Comment (thomasvl) [inline Tests/SwiftProtobufTests/Test_Map.swift]: Updated the second commit to disable this assert and opened #334 to make sure we have this right.

**Pull request merged**